File: packages/api/app/db/schema.py
"""MongoDB collection and index definitions for jackson-api.

This artifact is intentionally committed even though local tests use an in-memory
repository. It is the deployment schema contract for MongoDB-backed storage.
It stores token hashes only; there is no plaintext token field.
"""

COLLECTIONS = {
    "users": {
        "fields": {
            "_id": "ObjectId | string",
            "username": "string",
            "username_normalized": "string",
            "token_hash": "sha256 hex string",
            "created_at": "datetime",
        },
        "indexes": [
            {"keys": [("username_normalized", 1)], "unique": True, "name": "uniq_username_normalized"},
            {"keys": [("token_hash", 1)], "unique": True, "name": "uniq_token_hash"},
        ],
    },
    "requests": {
        "fields": {
            "_id": "request_id string",
            "user_id": "users._id",
            "message": "string",
            "status": "pending | replied",
            "created_at": "datetime",
            "updated_at": "datetime",
        },
        "indexes": [
            {"keys": [("user_id", 1), ("created_at", -1)], "name": "by_user_created"},
        ],
    },
    "replies": {
        "fields": {
            "request_id": "requests._id",
            "reply_text": "string",
            "telegram_update_id": "int | null",
            "telegram_message_id": "int | null",
            "created_at": "datetime",
        },
        "indexes": [
            {"keys": [("request_id", 1)], "unique": True, "name": "uniq_reply_per_request"},
            {"keys": [("telegram_update_id", 1)], "unique": True, "sparse": True, "name": "uniq_reply_update_id"},
        ],
    },
    "telegram_messages": {
        "fields": {
            "request_id": "requests._id",
            "operator_chat_id": "int",
            "sent_message_id": "int",
            "created_at": "datetime",
        },
        "indexes": [
            {"keys": [("request_id", 1)], "unique": True, "name": "uniq_telegram_message_request"},
            {"keys": [("operator_chat_id", 1), ("sent_message_id", 1)], "unique": True, "name": "uniq_operator_sent_message"},
        ],
    },
    "telegram_webhook_updates": {
        "fields": {
            "update_id": "int",
            "status": "processed | ignored | duplicate",
            "request_id": "requests._id | null",
            "message_id": "int | null",
            "created_at": "datetime",
        },
        "indexes": [
            {"keys": [("update_id", 1)], "unique": True, "name": "uniq_telegram_update_id"},
            {"keys": [("request_id", 1), ("status", 1)], "name": "by_request_status"},
        ],
    },
}


def collection_names() -> list[str]:
    return list(COLLECTIONS.keys())


def indexes_for(collection_name: str) -> list[dict]:
    return COLLECTIONS[collection_name]["indexes"]

File: apps/api/src/db/schema.ts
import {
  bigint,
  char,
  index,
  pgSchema,
  primaryKey,
  text,
  timestamp,
  uniqueIndex,
  uuid,
} from 'drizzle-orm/pg-core'

export const jacksonSchema = pgSchema('jackson')

export const requestStatus = jacksonSchema.enum('request_status', ['pending', 'replied'])
export const webhookUpdateStatus = jacksonSchema.enum('webhook_update_status', [
  'processed',
  'ignored',
  'duplicate',
])

const timestampColumn = (name: string) =>
  timestamp(name, { mode: 'date', withTimezone: true }).defaultNow().notNull()

export const users = jacksonSchema
  .table(
    'users',
    {
      id: uuid('id').defaultRandom().primaryKey(),
      username: text('username').notNull(),
      usernameNormalized: text('username_normalized').notNull(),
      tokenHash: char('token_hash', { length: 64 }).notNull(),
      createdAt: timestampColumn('created_at'),
    },
    (table) => [
      uniqueIndex('uniq_username_normalized').on(table.usernameNormalized),
      uniqueIndex('uniq_token_hash').on(table.tokenHash),
    ],
  )
  .enableRLS()

export const requests = jacksonSchema
  .table(
    'requests',
    {
      id: text('id').primaryKey(),
      userId: uuid('user_id')
        .notNull()
        .references(() => users.id),
      message: text('message').notNull(),
      status: requestStatus('status').default('pending').notNull(),
      createdAt: timestampColumn('created_at'),
      updatedAt: timestampColumn('updated_at'),
    },
    (table) => [index('by_user_created').on(table.userId, table.createdAt.desc())],
  )
  .enableRLS()

export const replies = jacksonSchema
  .table(
    'replies',
    {
      requestId: text('request_id')
        .notNull()
        .references(() => requests.id),
      replyText: text('reply_text').notNull(),
      telegramUpdateId: bigint('telegram_update_id', { mode: 'number' }),
      telegramMessageId: bigint('telegram_message_id', { mode: 'number' }),
      createdAt: timestampColumn('created_at'),
    },
    (table) => [
      primaryKey({ name: 'uniq_reply_per_request', columns: [table.requestId] }),
      uniqueIndex('uniq_reply_update_id').on(table.telegramUpdateId),
    ],
  )
  .enableRLS()

export const telegramMessages = jacksonSchema
  .table(
    'telegram_messages',
    {
      requestId: text('request_id')
        .notNull()
        .references(() => requests.id),
      operatorChatId: bigint('operator_chat_id', { mode: 'number' }).notNull(),
      sentMessageId: bigint('sent_message_id', { mode: 'number' }).notNull(),
      createdAt: timestampColumn('created_at'),
    },
    (table) => [
      primaryKey({ name: 'uniq_telegram_message_request', columns: [table.requestId] }),
      uniqueIndex('uniq_operator_sent_message').on(table.operatorChatId, table.sentMessageId),
    ],
  )
  .enableRLS()

export const telegramWebhookUpdates = jacksonSchema
  .table(
    'telegram_webhook_updates',
    {
      updateId: bigint('update_id', { mode: 'number' }).notNull(),
      status: webhookUpdateStatus('status').notNull(),
      requestId: text('request_id').references(() => requests.id),
      messageId: bigint('message_id', { mode: 'number' }),
      createdAt: timestampColumn('created_at'),
    },
    (table) => [
      primaryKey({ name: 'uniq_telegram_update_id', columns: [table.updateId] }),
      index('by_request_status').on(table.requestId, table.status),
    ],
  )
  .enableRLS()
